src/bam/mod.rs

```diff
@@ -403,6 +403,157 @@ impl Drop for Writer {
+    f: *mut htslib::htsFile,
+    header: HeaderView,
+}
+
+/// Wrapper for opening a SAM file.
+fn hts_open(path: &ffi::CStr, mode: &[u8]) -> Result<*mut htslib::htsFile, SAMError> {
+    let ret = unsafe {
+        htslib::hts_open(
+            path.as_ptr(),
+            ffi::CString::new(mode).unwrap().as_ptr()
+        )
+    };
+    if ret.is_null() {
+        Err(SAMError::IOError)
+    } else {
+        Ok(ret)
+    }
+}
+
+
+impl SAMWriter {
+    /// Create new SAM file writer.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path.
+    /// * `header` - header definition to use
+    pub fn from_path<P: AsRef<Path>>(path: P, header: &header::Header) -> Result<Self, SAMError> {
+        if let Some(p) = path.as_ref().to_str() {
+            Ok(try!(Self::new(p.as_bytes(), header)))
+        } else {
+            Err(SAMError::IOError)
+        }
+    }
+
+    /// Create a new SAM file at STDOUT.
+    ///
+    /// # Arguments
+    ///
+    /// * `header` - header definition to use
+    pub fn from_stdout(header: &header::Header) -> Result<Self, SAMError> {
+        Self::new(b"-", header)
+    }
+
+    fn new(path: &[u8], header: &header::Header) -> Result<Self, SAMError> {
+        let f = try!(hts_open(&ffi::CString::new(path).unwrap(), b"w"));
+        let header_record = unsafe {
+            let header_string = header.to_bytes();
+            let l_text = header_string.len();
+            let text = ::libc::malloc(l_text + 1);
+            ::libc::memset(text, 0, l_text + 1);
+            ::libc::memcpy(text, header_string.as_ptr() as *const ::libc::c_void, header_string.len());
+            //println!("{}", std::str::from_utf8(&header_string).unwrap());
+            let rec = htslib::sam_hdr_parse(
+                (l_text + 1) as i32,
+                text as *const i8,
+            );
+            (*rec).text = text as *mut i8;
+            (*rec).l_text = l_text as u32;
+            rec
+        };
+        unsafe { htslib::sam_hdr_write(f, header_record); }
+        Ok(SAMWriter { f: f, header: HeaderView::new(header_record) })
+    }
+
+    /// Write record to SAM.
+    ///
+    /// # Arguments
+    ///
+    /// * `record` - the record to write
+    pub fn write(&mut self, record: &record::Record) -> Result<(), WriteError> {
+        if unsafe { htslib::sam_write1(self.f, &self.header.inner(), record.inner) } == -1 {
+            Err(WriteError::Some)
+        }
+        else {
+            Ok(())
+        }
+    }
+
+    /// Read bam file. For each record apply f to it, and write to sam file if f returned Some(true), skip record if Some(false) if None then terminate iteration
+    ///
+    /// # Arguments
+    ///
+    /// * `bamfile` - the bam file to read from
+    /// * `samfile` - the sam file to write
+    /// * `f` - the predicate to apply
+    pub fn from_bam_with_filter<'a, 'b, F>(bamfile:&'a str, samfile:&'b str, f:F) -> Result<(), SAMError> where F:Fn(&record::Record) -> Option<bool> {
```

Comment:
I thought so too. I was using that function for the SAMWriter test, I will make it private with the there.